c2corg_api/views/document_delete.py

```diff
@@ -0,0 +1,327 @@
+        filter(and_(
+            Association.parent_document_id == document_id,
+            Association.child_document_type == ROUTE_TYPE,
+        )).subquery()
+    only_waypoint = DBSession.query(Association). \
+        filter(and_(
+            Association.child_document_id == routes.c.child_document_id,
+            Association.parent_document_type == WAYPOINT_TYPE
+        )). \
+        group_by(Association.child_document_id). \
+        having(func.count('*') == 1). \
+        exists()
+    return DBSession.query(only_waypoint).scalar()
+
+
+def _is_only_route_of_outing(document_id):
+    outings = DBSession.query(Association.child_document_id). \
+        filter(and_(
+            Association.parent_document_id == document_id,
+            Association.child_document_type == OUTING_TYPE,
+        )).subquery()
+    only_route = DBSession.query(Association). \
+        filter(and_(
+            Association.child_document_id == outings.c.child_document_id,
+            Association.parent_document_type == ROUTE_TYPE
+        )). \
+        group_by(Association.child_document_id). \
+        having(func.count('*') == 1). \
+        exists()
+    return DBSession.query(only_route).scalar()
+
+
+@resource(path='/documents/delete/{id}', cors_policy=cors_policy)
+class DeleteDocumentRest(object):
+
+    def __init__(self, request):
+        self.request = request
+
+    @restricted_json_view(
+        permission='moderator',
+        validators=[validate_id, validate_document])
+    def delete(self):
+        """
+        Delete a document.
+
+        Request:
+            `DELETE` `/documents/delete/{id}`
+        """
+        document_id = self.request.validated['id']
+        document_type = self.request.validated['document_type']
+
+        # Note: if an error occurs while deleting, SqlAlchemy will
+        # automatically cancel all DB changes.
+
+        if document_type == IMAGE_TYPE:
+            # Files are actually removed only if the transaction succeeds
+            delete_all_files_for_image(document_id, self.request)
+
+        _remove_from_cache(document_id)
+
+        # Remove associations and update cache of formerly associated docs
+        update_cache_version_full(document_id, document_type)
+        _remove_associations(document_id)
+
+        clazz, clazz_locale, archive_clazz, archive_clazz_locale = _get_models(
+            document_type)
+
+        # Order of removals depends on foreign key constraints
+        _remove_history_metadata(document_id)
+        _remove_archive_locale(archive_clazz_locale, document_id)
+        _remove_archive_geometry(document_id)
+        _remove_archive(archive_clazz, document_id)
+        _remove_locale(clazz_locale, document_id)
+        _remove_geometry(document_id)
+        _remove_figures(clazz, document_id)
+        _remove_from_feed(document_id)
+
+        if document_type == IMAGE_TYPE:
+            # Remove this image references from the feed
```

Comment:
Yes, in feed entries you have references to images, see https://github.com/c2corg/v6_api/blob/master/c2corg_api/models/feed.py#L152-L163

If you delete one of these images and don't update the feed entries that reference it, you'll get broken images/links in the remaining feed cards. I have tried to address this problem with https://github.com/c2corg/v6_api/blob/78e8fdc0445213a283a4eedf136a56afa0054369/c2corg_api/views/document_delete.py#L284-L305